### PR TITLE
Fixed Maxbounds, Snap, and added Undo.

### PR DESCRIPTION
**Pull request by germanjoey**
Hello,

This is the L.Draw half of a compound PR I'm sending to L.Draw and
L.Snap. The changes to L.Snap require the changes to L.Draw, but the
changes to L.Draw does not require the changes to L.Snap.

Notes:
- I don't have a touch-capable device available until at least
Christmas, so if anyone can help check these changes w/ touch that would
be great.
- I folded in the changes to L.Snap in Leaflet-Draw/docs/examples/libs,
so you can see all this stuff as
Leaflet-Draw/docs/examples/crssimple_snapping.html
- I had intended to add write some real documentation for all this, but
this kiddo just won't sleep and I already had to merge all the changes
after 0.4.3 by hand, so I figure I better send this in now before I get
even more out of sync. I will send another PR with some actual words if
there's interest in this PR.

Changelist:

L.Draw
- Bugfix that forces L.Draw to respect a map's maxbounds
- very important for using L.Draw with L.CRS.Simple
- only does anything if map.options.maxbounds is not null
- new drawn layers (for all shape types) must be inside the max bounds
- edited layers (for both move and resize, for all shape types) are
forced to stay inside the maxbounds
- Some bugfixes to allow L.Snap to work in editmode with Leaflet 1.0
- Various L.Draw events now pass much more information in the event
object. (needed for undo/redo)
- Undo/Redo implemented as L.Draw.UndoManager
- works with all draw and edit shape types
- bound to ctrl-z/ctrl-y by default, but customizable
- Bugfixed Poly self-intersection logic for draw and edit

L.Snap
- Various bugfixes to allow it to work with L.Draw for Leaflet 1.0
- Various bugfixes in L.Snap logic
- Snap now works for all shape types (i.e. polymove, rect, circle, and
marker, not just polyline/polygon)
- a new L.Draw.Guidelines class, which extends orthogonal lines to all
sides of a bounding box during draw/edit for all rectangles and circles
in drawnLayers
- requires a bounding box, either a maxbounds on the map or passed in
options
- works with snap
- a new L.Draw.Gridlines class, which draws a grid
- requires a bounding box, either a maxbounds on the map or passed in
options
- can be set by either describing spacing or number of lines in the box,
in either latlng or pixels
- works with snap
- snap will always prefer to snap to drawn features (or their vertices,
if snapToVerticies is enabled) over Guidelines or Gridlines

### CODE REVIEW

**Comment by ddproxy** [comment]
I see breaking things.

You cool if I pull and piecemeal this?

**Comment by germanjoey** [comment]
Uh, sure as long as it still works with the PR I sent to L.Snap.

What's breaking?

**Comment by ddproxy** [comment]
Sorry, not quite 'breaking' as much as there are conflicts. I'd like to separate the changes against the most recent build (which needs to be patched, desperately).

Also did you intend on adding the images to the commit?

**Comment by germanjoey** [comment]
Hmmm, I thought I had fully patched by hand against the latest version of
L.Draw that was on github. Apologies for the trouble.

The images - yep. Their used by the new example, at
docs/examples/crssimple_snapping.html
This file demonstrates using L.Draw with L.CRS.Simple as well as the new
updates to snap. (all draw types are now working with snap, gridlines,
guidelines - leaflet.snap.js in docs/examples/libs has been updated with
this new version; still haven't heard from the snap guys about the PR I
sent to them.)

**Comment by germanjoey** [comment]
Hey,

I finally got around to updating the docs. Want me to send another PR, or..?

Joe

**Comment by ddproxy** [comment]
@germanjoey Is this in a branch on your fork? I can pull directly from it.

**Comment by ddproxy** [comment]
Hey @germanjoey !

So, review https://rawgit.com/germanjoey/Leaflet.draw/master/docs/examples/full.html which should be running your fork. The branch breaks normal methods (outside of just snap).

Additionally you have committed .bak files by accident and the tests don't pass (`npm run test` or `jake test`). If you are using an IDE like intellij, you can wire the test suite to run the tests directly with Karma so you can get feedback immediately after making changes.

There are several bugs hidden within the PR, one is using `map` when it's not available in the function scope.

While I can't take the PR in full, I appreciate the work done here and would like to work with you to resolve these issues.

I'll check what I can do to get the Snap plugin either pulled into the Leaflet organization or better visibility into getting updates pushed that direction.

I've pushed a branch to https://github.com/Leaflet/Leaflet.draw/tree/snap-undo-bounds that has 0.4.7 rebased with your fork. There are a few fixes for the above issues but I hadn't completed them yet. I've also tweaked the package json and bower files to identify as 0.5.0 so I don't merge this on accident.

You can contact me directly at ddproxy@gmail.com, https://gitter.im/Leaflet/Leaflet.draw occasionally and on skype at ddproxy .

**Comment by germanjoey** [comment]
Yeah, just synced it.

**Comment by germanjoey** [comment]
Hey, thanks, I'll take a look at this later. What exactly do the tests do,
and what is failing in them?

**Comment by germanjoey** [comment]
Hey, I tried running the example at https://rawgit.com/germanjoey/
Leaflet.draw/master/docs/examples/full.html and it doesn't seem to do
anything at all. There's no snapping, no undo/redo, polygon drawing does
not work.  Could you explain a little more about what I'm supposed to see
here, or what you've brought in from my

The .bak files - sorry, that's my bad. My other projects all have that in
.gitignore and so I didn't even think about it.

Running "npm install" and then "jake test", both in my fork as well as the
unmodified master build, results in:

  Checking for js errors...
  jake aborted.
  const fs = require("fs"),
  ^^^^^
  SyntaxError: use of const in strict mode.

What else do I need to do to run these tests?

**Comment by germanjoey** [comment]
Oh, I see what you're saying now, regarding the issue with
https://rawgit.com/germanjoey/Leaflet.draw/master/docs/examples/full.html -
I did all my testing with snap enabled, and never thought to test without
snap, woops! Ok, lemme fix this up in the morning, and I'll let you know
when there's something in my fork you can pull.

Joe

**Comment by germanjoey** [comment]
Ok, I think everything except Polyline/Polygon drawing is now functioning
correctly on all the example files. The ones that were definitely my fault:

  1.) The error on Circle-Edit-Move and Polyline drawing not registering
clicks were due to me not testing enough with a map without a maxBounds and
having faulty if-guards; fixed easily enough.
  2.) The bound on map in UndoManager shoulda been this._map, woops.

However, the other stuff seems to be related to fault touch detection, from
some sort of commit that occurred between 0.4.4 (the version I actually
implemented my changes on) and 0.4.7. I admit that I could have noticed
this before sending my original PR, but I did not do more than cursory
testing after merging the 0.4.4->0.4.7 changes as I didn't expect anything
important to have changed. Now, however, L.Draw has started issuing events
for both "mouseup" as well as "touchstart" despite the fact that I am using
a mouse. I believe this is the root problem that I recall some have been
talking recently about regarding the fact that sometimes L.Draw will create
double-points when drawing a Polyline or a Polygon, sometimes with a small
offset if the mouse has moved a couple milimeters during the execution time
of the touchstart event handler.

I plugged this problem by patching in a check in TouchEvents._touchEvent
that bails if the pointerType is a mouse, to prevent the double event. This
seems pretty janky to me though, so if someone who is more familar with the
operation of touch events could look into it, that would be wonderful. I
also forced Polyline._onMouseUp to obey snapping if its event came through
there, as the 0.4.7 implementation bypasses e to check e.originalEvent, for
some reason. I'm guessing (since I don't have a touch-enabled device to
check with) that _onTouch does not check snap at all, although I did patch
snap (again) to also update e.originalEvent if it exists in the event
object on snap.

I've got to admit that I find this issue pretty annoying. I'm thrilled
about touch support, but not if it's gonna break mouse support. Has this
been discussed before? I think the solution is to detect touch in a more
accurate way.  touch if: a.) the browser is not capable of pointer events,
e.g., "if (!L.Browser.pointer)" -or- if the user has tried to manipulate
the map via touch during the current tool session. (or if the session was
opened via touch?) I'm not exactly sure how to implement that second
condition, however.

Joe

**Comment by ddproxy** [comment]
There is a patch for the touch events in another pr. So, yeah not your fault. We removed the fault detection then encountered issues when browsers updated (yay).

I'll review soon :) thanks

Jon West

Jon West

**Comment by germanjoey** [comment]
Just pushed another commit to my fork, fixing a few subtle bugs with the
add/remove vertex undo/redo code in edit mode that I happened to notice
during my earlier debugging session.

There's one more issue I discovered when debugging the touch stuff: if the
user repeatedly hits undo/redo very very fast - i.e. faster than the
browser can execute one item on the stack - then either Leaflet or the
javascript engine itself will mistakenly trigger a deep recursion error.
It's obviously not a real recursion error, but some sort of race condition
that occurs between two stack items attempting to manipulate the same state.

I think the solution is to simply add the undo and redo items to process
into a priority queue when undo is called, and then process items in that
queue until it is fully empty, blocking further processing calls in the
meantime. That way, subsequent calls to undo (triggered by keypresses) will
simply add more items to the queue, rather than attempt to execute them
independently in a way that could lead to a race condition.

Anyways, I'm gonna hold off on implementing this until all the current
changes are in. It shouldn't be too hard but it will make the code a little
harder to follow, and it's already pretty hard to follow, IMHO due to all
the state changes that are spread out across different event handlers.

Joe

**Comment by v1r0x** [comment]
Any news on this? Do you need any help (e.g. testing)? Would like to help to get this merged :wink:

**Comment by germanjoey** [comment]
I've updated the code in PR a lot on my end, new features and bugfixes, but
haven't pushed any of it because the main problem is still finding someone
on the project who will merge it in. =/ If we can find someone, another
person who could help with testing would be really great.

Joe

**Comment by v1r0x** [comment]
@germanjoey Could you push your new changes? We decided to fork this repo and leaflet.snap as well, because we really need the features. If this is ok for you, I try to help to fix conflicts and any errors. What do you think @germanjoey ?

**Comment by germanjoey** [comment]
Hi,

Yeah, I can do that. I'm in the middle of something else right now that
should take a day or two to finish up, but then after that I'll re-merge
what I have and send another push.

Joe

**Comment by ddproxy** [comment]
@germanjoey Please check https://github.com/Leaflet/Leaflet.draw/tree/undo-manager for differences/compatibility, etc. This `undo-manager` has this PR and L.Draw 0.4.10 merged, code styles updated, etc. There's a bit of cleanup to be done in documentation and tests that should happen against that branch.

**Comment by germanjoey** [comment]
Pushed a commit to Leaflet.draw-undo-manager, fixing some stuff broken in
the merge.

docs/examples/crs_simple.html works for me, although the toolbar icons seem
to be all screwed up for some reason. I didn't look into that too much,
just made sure undo and snap were working. I also didn't think to look
whether the leaflet.js in docs/examples/libs was leaflet 1.1 or not, just
thought about that now.

What documentation/test stuff is needed?

Joe

**Comment by ddproxy** [comment]
Checkout the undo-manager branch and run `jake` for tests. There are a few failing. I'm also having problems drawing, anything, in the examples.

**Comment by germanjoey** [comment]
Which examples, specifically? Were you able to see my most recent commit to
the undo-manager branch? (and did I actually push the commit to that
branch? I have to admit, that I'm not used to working with branches w/
github).

I'll take a look at the jake tests.

**Comment by ddproxy** [comment]
I don't see it, so I'm not sure it actually happened... Git should have told you if it was unsuccessful. You can dump the git command and response here if you'd like.

**Comment by germanjoey** [comment]
Alright, specs and examples are fixed now. Looks like the screwy icons were
because the .sr-only tag was missing from leaflet.draw.css - might wanna
look into that, since IIRC I've never touched that file. Re-added it for
now.

Anyways, here's the two pushes:

https://github.com/germanjoey/Leaflet.draw/commit/28a5209749182d51f069449bcc81827749cb4b3e

https://github.com/germanjoey/Leaflet.draw/commit/094bea552e1adfebfc1f4d4e18c1c11078297351

Did it send these to the fork under my account instead of
leaflet.draw-undo-manager? Why would I do that when I started from a clean
clone of your repo? Can you transfer these over in some way?

**Comment by germanjoey** [comment]
So, should I try pushing again here?

**Comment by ddproxy** [comment]
@germanjoey yes please, looking at this in the next 5 hours.

**Comment by germanjoey** [comment]
It work this time? I think so?

https://github.com/Leaflet/Leaflet.draw/pull/745

**Comment by ddproxy** [comment]
@germanjoey I'm closing this PR in favor of working against the branch. Will have time to look at that a bit this evening and weekend.

**Comment by v1r0x** [comment]
Sorry for commenting this old PR, but I wanted to ask status of the undo-manager branch? Didn't see any progress in a while. Is there something I could do to help?
